pkg/controller/azvolumeattachment.go

```diff
@@ -451,20 +460,71 @@ func (r *reconcileAzVolumeAttachment) DeleteFinalizer(ctx context.Context, attac
 	return nil
 }
 
-func finalizerExists(azVolumeAttachment v1alpha1.AzVolumeAttachment, finalizerName string) bool {
-	if azVolumeAttachment.Finalizers != nil {
-		for _, finalizer := range azVolumeAttachment.Finalizers {
-			if finalizer == finalizerName {
-				return true
-			}
+func (r *reconcileAzVolumeAttachment) AddFinalizerToAzVolume(ctx context.Context, volumeName string) error {
+	var azVolume v1alpha1.AzVolume
+	if err := r.client.Get(ctx, types.NamespacedName{Namespace: r.namespace, Name: volumeName}, &azVolume); err != nil {
+		klog.Errorf("failed to get AzVolume (%s): %v", volumeName, err)
+		return err
+	}
+
+	updated := azVolume.DeepCopy()
+
+	if updated.Finalizers == nil {
+		updated.Finalizers = []string{}
+	}
+
+	if finalizerExists(updated.Finalizers, AzVolumeAttachmentFinalizer) {
+		return nil
+	}
+
+	updated.Finalizers = append(updated.Finalizers, AzVolumeAttachmentFinalizer)
+	if err := r.client.Update(ctx, updated, &client.UpdateOptions{}); err != nil {
+		klog.Errorf("failed to add finalizer (%s) to AzVolume(%s): %v", AzVolumeAttachmentFinalizer, updated.Name, err)
+		return err
+	}
+	klog.Infof("successfully added finalizer (%s) to AzVolume (%s)", AzVolumeAttachmentFinalizer, updated.Name)
+	return nil
+}
+
+func (r *reconcileAzVolumeAttachment) DeleteFinalizerFromAzVolume(ctx context.Context, volumeName string) error {
+	var azVolume v1alpha1.AzVolume
+	if err := r.client.Get(ctx, types.NamespacedName{Namespace: r.namespace, Name: volumeName}, &azVolume); err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		klog.Errorf("failed to get AzVolume (%s): %v", volumeName, err)
+		return err
+	}
+
+	updated := azVolume.DeepCopy()
+
+	for _, finalizer := range updated.Finalizers {
+		if finalizer == AzVolumeAttachmentFinalizer {
+			continue
+		}
+		updated.Finalizers = append(updated.Finalizers, finalizer)
```

Comment:
Here all the finalizers except AzVolumeAttachmentFinalizer  are getting appended to the updated.Finalizers list. I think this will not remove the AzVolumeAttachmentFinalizer   finalizer from  updated.Finalizers.

Reply:
ay, you are definitely right. I must have forgotten to initialize a new slice. Finalizers need to be appended to a new slice and assigned to `updated.Finalizers`.